File: packages/server/src/services/content-extractor.ts
import { Readability } from '@mozilla/readability'
import { parseHTML } from 'linkedom'
import type { ContentType } from '@clawssify/shared'

export interface ExtractedContent {
  title: string
  text: string
  url?: string
}

export async function extractContent(
  type: ContentType,
  content: string,
): Promise<ExtractedContent> {
  switch (type) {
    case 'url':
      return extractFromUrl(content)
    case 'tweet':
      if (isTweetUrl(content)) {
        return extractFromUrl(content.trim())
      }
      return extractFromText(content)
    case 'text':
    case 'note':
    case 'conversation':
      return extractFromText(content)
  }
}

function isTweetUrl(content: string): boolean {
  const trimmed = content.trim()
  return trimmed.startsWith('http') && (trimmed.includes('twitter.com') || trimmed.includes('x.com'))
}

function generateTitleFromText(text: string): string {
  const firstLine = text.split('\n')[0].trim()
  if (firstLine.length <= 60) return firstLine
  return `${firstLine.slice(0, 60)}...`
}

function extractFromText(content: string): ExtractedContent {
  return {
    title: generateTitleFromText(content),
    text: content,
  }
}

async function extractFromUrl(url: string): Promise<ExtractedContent> {
  const trimmedUrl = url.trim()

  try {
    const response = await fetch(trimmedUrl, {
      headers: { 'User-Agent': 'Clawssify/0.1 (Knowledge Base Bot)' },
      signal: AbortSignal.timeout(15000),
    })
    const html = await response.text()
    const { document } = parseHTML(html)
    const reader = new Readability(document)
    const article = reader.parse()

    if (article) {
      return {
        title: article.title,
        text: article.textContent,
        url: trimmedUrl,
      }
    }

    return {
      title: document.title || trimmedUrl,
      text: document.body?.textContent?.trim() || '',
      url: trimmedUrl,
    }
  } catch {
    return {
      title: trimmedUrl,
      text: `Failed to extract content from URL: ${trimmedUrl}`,
      url: trimmedUrl,
    }
  }
}

File: packages/core/src/engine/content_extractor.rs
//! URL readability + text extraction — port of `server/src/services/content-extractor.ts`.

use crate::models::{ContentType, ExtractedContent};
use crate::CoreError;

/// Extract content from the given input based on content type.
pub async fn extract_content(
    content_type: &ContentType,
    content: &str,
) -> Result<ExtractedContent, CoreError> {
    match content_type {
        ContentType::Url => extract_from_url(content).await,
        ContentType::Tweet => {
            if is_tweet_url(content) {
                extract_from_url(content.trim()).await
            } else {
                Ok(extract_from_text(content))
            }
        }
        ContentType::Text | ContentType::Note | ContentType::Conversation => {
            Ok(extract_from_text(content))
        }
    }
}

fn is_tweet_url(content: &str) -> bool {
    let trimmed = content.trim();
    trimmed.starts_with("http")
        && (trimmed.contains("twitter.com") || trimmed.contains("x.com"))
}

fn generate_title_from_text(text: &str) -> String {
    let first_line = text.lines().next().unwrap_or("").trim();
    if first_line.len() <= 60 {
        first_line.to_string()
    } else {
        format!("{}...", &first_line[..60])
    }
}

fn extract_from_text(content: &str) -> ExtractedContent {
    ExtractedContent {
        title: generate_title_from_text(content),
        text: content.to_string(),
        url: None,
    }
}

async fn extract_from_url(url: &str) -> Result<ExtractedContent, CoreError> {
    let trimmed_url = url.trim();

    let client = reqwest::Client::builder()
        .user_agent("Clawssify/0.1 (Knowledge Base Bot)")
        .timeout(std::time::Duration::from_secs(15))
        .build()
        .map_err(|e| CoreError::Network {
            msg: e.to_string(),
        })?;

    let response = client
        .get(trimmed_url)
        .send()
        .await
        .map_err(|e| CoreError::ContentExtraction {
            msg: e.to_string(),
        })?;

    let html = response
        .text()
        .await
        .map_err(|e| CoreError::ContentExtraction {
            msg: e.to_string(),
        })?;

    // Use scraper to extract readable text (similar to Readability)
    let document = scraper::Html::parse_document(&html);

    // Extract title
    let title_selector = scraper::Selector::parse("title").unwrap();
    let title = document
        .select(&title_selector)
        .next()
        .map(|el| el.text().collect::<String>())
        .unwrap_or_else(|| trimmed_url.to_string());

    // Extract body text — prefer <article>, fall back to <body>
    let text = extract_article_text(&document).unwrap_or_else(|| extract_body_text(&document));

    Ok(ExtractedContent {
        title,
        text,
        url: Some(trimmed_url.to_string()),
    })
}

fn extract_article_text(document: &scraper::Html) -> Option<String> {
    let article_selector = scraper::Selector::parse("article").ok()?;
    let article = document.select(&article_selector).next()?;
    let text: String = article.text().collect::<Vec<_>>().join(" ");
    let trimmed = text.trim().to_string();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed)
    }
}

fn extract_body_text(document: &scraper::Html) -> String {
    let body_selector = scraper::Selector::parse("body").unwrap();
    document
        .select(&body_selector)
        .next()
        .map(|el| el.text().collect::<Vec<_>>().join(" ").trim().to_string())
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_from_text() {
        let result = extract_from_text("Hello world\nSecond line");
        assert_eq!(result.title, "Hello world");
        assert_eq!(result.text, "Hello world\nSecond line");
        assert!(result.url.is_none());
    }

    #[test]
    fn test_title_truncation() {
        let long_text = "A".repeat(100);
        let result = extract_from_text(&long_text);
        assert!(result.title.ends_with("..."));
        assert_eq!(result.title.len(), 63); // 60 + "..."
    }

    #[test]
    fn test_is_tweet_url() {
        assert!(is_tweet_url("https://twitter.com/user/status/123"));
        assert!(is_tweet_url("https://x.com/user/status/123"));
        assert!(!is_tweet_url("Just a tweet about something"));
    }
}
